### PR TITLE
Add Support for Disabling Introspection Queries

### PR DESCRIPTION
**Pull request by DimuthuMadushan**
## Purpose

Fixes: [#2891](https://github.com/ballerina-platform/ballerina-standard-library/issues/2891)

## Examples
The GraphQL introspection query support can be enabled or disabled using service configurations as follows.

Disable introspection query support:
```ballerina
@graphql:ServiceConfig {
    introspection: false
}
service /introspection on new graphql:Listener(9000) {

    isolated resource function get name() returns string {
        return "Ballerina GraphQL";
    }
}
```

## Checklist
- [x] Linked to an issue
- [ ] Updated the changelog
- [x] Added tests
- [x] Updated the spec

### CODE REVIEW

File: ballerina/introspection_validator_visitor.bal [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        foreach parser:SelectionNode selection in operationNode.getSelections() {
+            selection.accept(self, operationNode.getKind());
+        }
+    }
+
+    public isolated function visitField(parser:FieldNode fieldNode, anydata data = ()) {
+        if fieldNode.getName() == SCHEMA_FIELD || fieldNode.getName() == TYPE_FIELD {
+            if !self.introspection {
+                string message = string `GraphQL introspection is not allowed by the GraphQL Service, but the query` +
+                                 string ` contained __schema or __type.`;
```
Comment (ThisaruGuruge) [inline ballerina/introspection_validator_visitor.bal:47]: We can make this more specific error, right?

File: ballerina/introspection_validator_visitor.bal [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+    public isolated function visitOperation(parser:OperationNode operationNode, anydata data = ()) {
+        foreach parser:SelectionNode selection in operationNode.getSelections() {
+            selection.accept(self, operationNode.getKind());
+        }
+    }
+
+    public isolated function visitField(parser:FieldNode fieldNode, anydata data = ()) {
+        if fieldNode.getName() == SCHEMA_FIELD || fieldNode.getName() == TYPE_FIELD {
+            if !self.introspection {
```
Comment (ThisaruGuruge) [inline ballerina/introspection_validator_visitor.bal:45]: We can just return from the `visitDocument` function if the introspection is disabled. No need to iterate through everything here.

File: ballerina/engine.bal [hunk 3]
```diff
@@ -101,7 +103,8 @@ isolated class Engine {
             new VariableValidatorVisitor(self.schema, variables),
             new FieldValidatorVisitor(self.schema),
             new DirectiveValidatorVisitor(self.schema),
-            new SubscriptionValidatorVisitor()
+            new SubscriptionValidatorVisitor(),
+            new IntrospectionValidatorVisitor(self.introspection)
         ];
```
Comment (ThisaruGuruge) [inline ballerina/engine.bal:108]:
```suggestion
            new SubscriptionValidatorVisitor()
        ];
        if self.enableIntrospection {
            validators.push(new IntrospectionValidatorVisitor());
        }
```

File: ballerina/engine.bal [hunk 4]
```diff
@@ -22,15 +22,17 @@ isolated class Engine {
     private final readonly & __Schema schema;
     private final int? maxQueryDepth;
     private final readonly & (readonly & Interceptor)[] interceptors;
+    private final readonly & boolean introspection;
```
Comment (ThisaruGuruge) [inline ballerina/engine.bal:25]:
```suggestion
    private final readonly & boolean enableIntrospection;
```

File: ballerina/annotations.bal [hunk 5]
```diff
@@ -23,6 +23,7 @@
 # + graphiql - GraphiQL client configurations
 # + schemaString - The generated schema. This is auto-generated at the compile-time
 # + interceptors - GraphQL service level interceptors
+# + enableIntrospection - The state of the introspection query support
```
Comment (ThisaruGuruge) [inline ballerina/annotations.bal:26]:
```suggestion
# + enableIntrospection - Whether to enable or disable the introspection on the service
```

File: ballerina/engine.bal [hunk 6]
```diff
@@ -22,15 +22,17 @@ isolated class Engine {
     private final readonly & __Schema schema;
     private final int? maxQueryDepth;
     private final readonly & (readonly & Interceptor)[] interceptors;
+    private final readonly & boolean enableIntrospection;
 
-    isolated function init(string schemaString, int? maxQueryDepth, Service s, 
-                           readonly & (readonly & Interceptor)[] interceptors) returns Error? {
+    isolated function init(string schemaString, int? maxQueryDepth, Service s,
+                           readonly & (readonly & Interceptor)[] interceptors, boolean enableIntrospection) returns Error? {
```
Comment (ThisaruGuruge) [inline ballerina/engine.bal:28]: Does the line exceed the char limit?

File: docs/spec/spec.md [hunk 7]
```diff
@@ -1327,6 +1328,22 @@ service on new graphql:Listener(4000) {
 }
 ```
 
+#### 9.1.7 Introspection Configurations
+
+The `enableIntrospection` field is used to enable or disable the GraphQL `Introspection` query support. If the introspection query support is disabled, the GraphQL service won't allow to execute `__schema` and `__type` introspection queries. However, the `__typename` introspection is working even if the introspection query support is disabled.
```
Comment (ThisaruGuruge) [inline docs/spec/spec.md:1333]:
```suggestion
The `enableIntrospection` field is used to enable or disable the GraphQL introspection query support. If the introspection query support is disabled, the GraphQL service won't allow the execution of the `__schema` and the `__type` introspection queries. However, the `__typename` introspection will work even if the introspection query support is disabled.

> **Note:** By default, a Ballerina GraphQL service will enable introspection query support. It is recommended to disable introspection in production environments.
```

**Review by ThisaruGuruge — APPROVED** [review]

**Comment by shafreenAnfar** [comment]
We decided to change the config to `instrospectionEnabled`

**Review by MohamedSabthar — APPROVED** [review]

File: docs/spec/spec.md [hunk 8]
```diff
@@ -1327,6 +1328,21 @@ service on new graphql:Listener(4000) {
 }
 ```
 
+#### 9.1.7 Introspection Configurations
+
+The `introspectionEnabled` field is used to enable or disable the GraphQL introspection query support. If the introspection query support is disabled, the GraphQL service won't allow the execution of the `__schema` and the `__type` introspection queries. However, the `__typename` introspection will work even if the introspection query support is disabled.
+###### Example: Disable Introspection Query Support
```
Comment (ThisaruGuruge) [inline docs/spec/spec.md:1334]:
```suggestion

###### Example: Disable Introspection Query Support
```

File: ballerina/introspection_validator_visitor.bal [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import graphql.parser;
+
+class IntrospectionValidatorVisitor {
```
Comment (ThisaruGuruge) [inline ballerina/introspection_validator_visitor.bal:19]: I think we should move all the introspection-related validations inside this visitor. It will make the code more readable and may reduce the load of the field validator visitor. We can do it in a separate PR though. Shall we create an issue to track this?

Comment (DimuthuMadushan) [reply ballerina/introspection_validator_visitor.bal:19]: Tracked via [ballerina-standard-library#3348](https://github.com/ballerina-platform/ballerina-standard-library/issues/3348)

File: docs/spec/spec.md (hunk 8 above)
Comment (DimuthuMadushan) [reply docs/spec/spec.md:1334]: Fixed via https://github.com/ballerina-platform/module-ballerina-graphql/pull/939/commits/00766689c933cd8e23fb2a3a01ac4b03fe7f2fc3
